genDefunSymbols doesn't look into local declarations

Despite `genDefunSymbols` inhabiting a `DsMonad` context, it is oblivious to any local declarations. I discovered this recently when trying to write code that looks something like:

```haskell
{-# LANGUAGE TemplateHaskell #-}
module Bug where

import Data.Singletons.TH
import Language.Haskell.TH
import Language.Haskell.TH.Desugar

$(do synName <- newName "Id"
     a       <- newName "a"
     let syn = TySynD synName [PlainTV a] (VarT a)
     defuns <- withLocalDeclarations [syn] $
               genDefunSymbols [synName]
     pure $ syn:defuns)
```

This will fail with:

```
$ /opt/ghc/8.4.3/bin/ghc Bug.hs
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )

Bug.hs:8:3: error:
    Looking up Id_6989586621679034162 in the list of available declarations failed.
This lookup fails if the declaration referenced was made in the same Template
Haskell splice as the use of the declaration. If this is the case, put
the reference to the declaration in a new splice.
  |
8 | $(do synName <- newName "Id"
  |   ^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

The culprit lies in the definition of `genDefunSymbols`:

https://github.com/goldfirere/singletons/blob/fdd464f9b3a0f0cfe24e1e028dd35193fc4758c2/src/Data/Singletons/Promote.hs#L103-L108

Note that it uses `reifyWithWarning`, which simply fails if `qReify` fails (and does not attempt to look into local declarations). I think this could be easily fixed by replacing `reifyWithWarning` with `reifyWithLocals`.

--- a/src/Data/Singletons/CustomStar.hs
+++ b/src/Data/Singletons/CustomStar.hs
@@ -104,7 +104,7 @@ singletonStar names = do
   where -- get the kinds of the arguments to the tycon with the given name
         getKind :: DsMonad q => Name -> q [DKind]
         getKind name = do
-          info <- reifyWithWarning name
+          info <- reifyWithLocals name
           dinfo <- dsInfo info
           case dinfo of
             DTyConI (DDataD _ (_:_) _ _ _ _ _) _ ->

--- a/src/Data/Singletons/Promote.hs
+++ b/src/Data/Singletons/Promote.hs
@@ -46,7 +46,7 @@ import qualified GHC.LanguageExtensions.Type as LangExt
 genPromotions :: DsMonad q => [Name] -> q [Dec]
 genPromotions names = do
   checkForRep names
-  infos <- mapM reifyWithWarning names
+  infos <- mapM reifyWithLocals names
   dinfos <- mapM dsInfo infos
   ddecs <- promoteM_ [] $ mapM_ promoteInfo dinfos
   return $ decsToTH ddecs
@@ -103,7 +103,7 @@ promoteOnly qdec = do
 genDefunSymbols :: DsMonad q => [Name] -> q [Dec]
 genDefunSymbols names = do
   checkForRep names
-  infos <- mapM (dsInfo <=< reifyWithWarning) names
+  infos <- mapM (dsInfo <=< reifyWithLocals) names
   decs <- promoteMDecs [] $ concatMapM defunInfo infos
   return $ decsToTH decs
 

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -86,7 +86,7 @@ contract constructors. This is the point of buildDataLets.
 genSingletons :: DsMonad q => [Name] -> q [Dec]
 genSingletons names = do
   checkForRep names
-  ddecs <- concatMapM (singInfo <=< dsInfo <=< reifyWithWarning) names
+  ddecs <- concatMapM (singInfo <=< dsInfo <=< reifyWithLocals) names
   return $ decsToTH ddecs
 
 -- | Make promoted and singleton versions of all declarations given, retaining

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -98,6 +98,7 @@ tests =
     , compileAndDumpStdTest "NatSymbolReflexive"
     , compileAndDumpStdTest "T323"
     , compileAndDumpStdTest "T332"
+    , compileAndDumpStdTest "T342"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/tests/compile-and-dump/Singletons/T342.ghc84.template
+++ b/tests/compile-and-dump/Singletons/T342.ghc84.template
@@ -1,0 +1,18 @@
+Singletons/T342.hs:(0,0)-(0,0): Splicing declarations
+    do synName <- newName "MyId"
+       a <- newName "a"
+       let syn = TySynD synName [PlainTV a] (VarT a)
+       defuns <- withLocalDeclarations [syn] $ genDefunSymbols [synName]
+       pure $ syn : defuns
+  ======>
+    type MyId a = a
+    type MyIdSym1 a0123456789876543210 = MyId a0123456789876543210
+    instance SuppressUnusedWarnings MyIdSym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) MyIdSym0KindInference) GHC.Tuple.())
+    data MyIdSym0 a0123456789876543210
+      where
+        MyIdSym0KindInference :: forall a0123456789876543210 arg.
+                                 SameKind (Apply MyIdSym0 arg) (MyIdSym1 arg) =>
+                                 MyIdSym0 a0123456789876543210
+    type instance Apply MyIdSym0 a0123456789876543210 = MyId a0123456789876543210

--- a/tests/compile-and-dump/Singletons/T342.hs
+++ b/tests/compile-and-dump/Singletons/T342.hs
@@ -1,0 +1,12 @@
+module T342 where
+
+import Data.Singletons.TH
+import Language.Haskell.TH
+import Language.Haskell.TH.Desugar
+
+$(do synName <- newName "MyId"
+     a       <- newName "a"
+     let syn = TySynD synName [PlainTV a] (VarT a)
+     defuns <- withLocalDeclarations [syn] $
+               genDefunSymbols [synName]
+     pure $ syn:defuns)

